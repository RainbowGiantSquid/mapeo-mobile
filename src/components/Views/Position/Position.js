// @flow
import React from 'react';
import { StyleSheet, Text, TouchableHighlight, View } from 'react-native';
import { NavigationActions, withNavigation } from 'react-navigation';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { DARK_GREY, LIGHT_GREY, MANGO } from '@lib/styles';

type Props = {
  navigation: NavigationActions
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: DARK_GREY
  },
  forward: {
    backgroundColor: MANGO,
    height: 50
  },
  header: {
    flexDirection: 'row',
    height: 100
  },
  title: {
    fontSize: 20,
    fontWeight: '700',
    color: LIGHT_GREY
  }
});

const Position = (props: Props) => (
  <View style={styles.container}>
    <View style={styles.header}>
      <TouchableHighlight onPress={() => props.navigation.goBack()}>
        <Icon color="gray" name="close" size={25} />
      </TouchableHighlight>
      <Text style={styles.title}>Position</Text>
      <TouchableHighlight
        onPress={() => props.navigation.navigate('Categories')}
        style={styles.forward}
      >
        <Icon color="white" name="arrow-forward" size={25} />
      </TouchableHighlight>
    </View>
  </View>
  );

export default withNavigation(Position);
