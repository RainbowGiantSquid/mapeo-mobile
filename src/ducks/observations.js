// @flow
import { create } from '@src/lib/redux';
import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { createSelector } from '@src/lib/selector';
import { StoreState } from '@types/redux';
import { Observation } from '@types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer,
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      observations: { $set: keyBy(action.payload, 'id') },
    });

    return newState;
  },
});

export const selectContact = createSelector(
  [
    (state: StoreState, id: string): Observation => state.app.observations[id],
  ],
  (observation: Observation): Observation => observation,
  (observation: Observation, id: string): string => id,
);

export default [observationListReducer];
