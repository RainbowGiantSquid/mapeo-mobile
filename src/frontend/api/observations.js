import fixture from "./test_observations.json";

export default class ObservationsApi {
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
  }
  list(cb) {
    setTimeout(() => cb(null, fixture), 2000);
  }
  update() {}
  create() {}
}
