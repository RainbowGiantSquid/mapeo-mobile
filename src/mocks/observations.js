// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 'ObservationID',
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: 'Thu Apr 05 2018 23:06:45 GMT+0200 (CEST)',
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: null,

  ...observation
});
