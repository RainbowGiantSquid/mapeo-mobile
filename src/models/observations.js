// @flow
import type {
  Observation,
  ServerObservationCreate,
  ServerObservationResponse
} from '../types/observation';

export const createDefaultObservation = (): Observation => ({
  created_at: new Date().toISOString(),
  notes: '',
  categoryId: '',
  attachments: [],
  fields: []
});

// Properties at the top level of a server observation object
const SERVER_PROPS: Array<$Keys<ServerObservationResponse>> = [
  'id',
  'version',
  'created_at',
  'timestamp',
  'type',
  'lon',
  'lat',
  'schemaVersion',
  'ref',
  'metadata',
  'attachments',
  'fields',
  'tags'
];

// Properties at top level of local observation object copied from server obj
const LOCAL_PROPS: Array<$Keys<Observation>> = [
  'id',
  'version',
  'created_at',
  'lat',
  'lon',
  'fields'
];

export const parseObservationRequest = (
  observation: Observation
): ServerObservationCreate | ServerObservationResponse => {
  const serverObservation: Object = { tags: {} };

  // Maintain top-level props and copy other props as tags
  Object.keys(observation).forEach(prop => {
    if (SERVER_PROPS.includes(prop)) {
      serverObservation[prop] = observation[prop];
    } else {
      serverObservation.tags[prop] = observation[prop];
    }
  });

  // set field answers as individual tags
  observation.fields.forEach(field => {
    serverObservation.tags[field.name] = field.answer;
  });

  serverObservation.attachments = (observation.attachments || []).map(a => ({
    id: a
  }));

  serverObservation.type = 'observation';

  return serverObservation;
};

export const parseObservationResponse = (
  serverObservation: ServerObservationResponse
): Observation => {
  const obs = createDefaultObservation();

  Object.keys(serverObservation).forEach(prop => {
    if (LOCAL_PROPS.includes(prop)) {
      // $FlowFixMe
      obs[prop] = serverObservation[prop];
    }
  });

  obs.attachments = (serverObservation.attachments || []).map(a => a.id);

  const tags = serverObservation.tags || {};
  obs.notes = tags.notes || '';
  obs.categoryId = tags.categoryId || '';

  return obs;
};
