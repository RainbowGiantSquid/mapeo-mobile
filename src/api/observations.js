// @flow
import { jsonRequest } from './base';
import { keyBy } from 'lodash';
import { parseObservationResponse } from '../models/observations';
import querystring from 'querystring'

import type {
  Observation as ObservationType,
  ServerObservationResponse,
  ServerObservationCreate
} from '../types/observation';

class Observation {
  static list = (opts) =>
    jsonRequest({
      method: 'GET',
      route: '/observations?' + querystring.stringify(opts)
    }).map(observations => {
      var payload = observations.map(parseObservationResponse)
      return keyBy(payload, 'id')
    });

  static create = (observation: ServerObservationCreate) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: observation
    }).map(parseObservationResponse);

  static update = (observation: ServerObservationResponse) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: observation
    }).map(parseObservationResponse);
}

export default Observation;
