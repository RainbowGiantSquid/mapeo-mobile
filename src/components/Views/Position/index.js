// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';

import { observationUpdate } from '../../../ducks/observations';
import type { StoreState } from '../../../types/redux';
import Position from './Position';
import type { StateProps, DispatchProps } from './Position';

function mapStateToProps(state: StoreState): StateProps {
  return { selectedObservation: state.app.selectedObservation };
}

function mapDispatchToProps(dispatch: Dispatch): DispatchProps {
  return {
    updateObservation: observation => dispatch(observationUpdate(observation))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Position);
