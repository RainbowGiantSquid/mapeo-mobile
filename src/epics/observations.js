// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import { OBSERVATION_LIST, observationList } from '../ducks/observations';
import { Action } from '../types/redux';
import { Observation } from '../types/observation';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, Observation[]>>,
  store: any
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(
      action =>
        action.status === 'Start' &&
        store.getState().app &&
        store.getState().app.observations &&
        !store.getState().app.observations.length
    )
    .map(() =>
      observationList('', [
        {
          type: 'Old houses and farms',
          id: 1,
          lat: 0.106831547,
          lon: -77.42499473,
          link: 'link',
          created: new Date('2018-03-25T12:00:00Z'),
          name: 'Old houses and farms',
          notes: `Pedro and Maria's house`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/1-house.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Hunting sites',
          id: 2,
          lat: 0.088170932,
          lon: -77.44020885,
          link: 'link',
          created: new Date('2018-03-25T12:30:00Z'),
          name: 'Hunting sites',
          notes: `Don Pancho's hunting camp`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/4-camp.jpg')
            }
          ],
          icon: require('../images/categories/category_2.png'),
          mock: true
        },
        {
          type: 'Threat',
          id: 3,
          lat: 0.112411067,
          lon: -77.41378756,
          link: 'link',
          created: new Date('2018-02-03T17:30:00Z'),
          name: 'Threat',
          notes: `Oil spill 3 February 2018. Ruptured pipeline. Dead fish, oily water`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/5-threat.jpg')
            }
          ],
          icon: null,
          mock: true
        },
        {
          type: 'Medicinal plants',
          id: 4,
          lat: 0.09749025,
          lon: -77.43614404,
          link: 'link',
          created: new Date('2018-02-04T17:30:00Z'),
          name: 'Medicinal plants',
          notes: `Salola fruit plant - used in tea to heal inflamation`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/6-plant.jpg')
            }
          ],
          icon: require('../images/categories/category_6.png'),
          mock: true
        },
        {
          type: 'Waterfalls',
          id: 5,
          lat: 0.105818148,
          lon: -77.4408037,
          link: 'link',
          created: new Date('2018-03-04T17:30:00Z'),
          name: 'Waterfalls',
          notes: `Sacred waterfall, visited during vision quests for ancestors, for taking ayahuasca`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/7-waterfall.jpg')
            }
          ],
          icon: require('../images/categories/category_13.png'),
          mock: true
        },
        {
          type: 'Rivers and streams',
          id: 6,
          lat: 0.119004333,
          lon: -77.42374225,
          link: 'link',
          created: new Date('2018-03-24T17:30:00Z'),
          name: 'Rivers and streams',
          notes: `Point where pipeline crosses river`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/8-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Historical sites',
          id: 7,
          lat: 0.082856811,
          lon: -77.4429836,
          link: 'link',
          created: new Date('2018-03-24T12:30:00Z'),
          name: 'Historical sites',
          notes: `There used to be a village here in the 1940s, it was abandoned after an epidemic of flu which killed most of the elders and children.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/9-specialsite.jpg')
            }
          ],
          icon: require('../images/categories/category_17.png'),
          mock: true
        },
        {
          type: 'Gold',
          id: 8,
          lat: 0.108605553,
          lon: -77.41560494,
          link: 'link',
          created: new Date('2018-03-11T17:30:00Z'),
          name: 'Gold',
          notes: `Illegal gold miners 11 March 2018, excavators, dredges, open tailing pond, mercury.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/10-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_8.png'),
          mock: true
        }
      ])
    );

export default [observationListEpic];
