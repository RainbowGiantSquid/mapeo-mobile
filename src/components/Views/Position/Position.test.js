// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import Position from './Position';
import { createObservation } from '../../../mocks/observations';

describe('Position tests', () => {
  test('snapshots', () => {
    const observation = createObservation();
    const isFocused = () => true;
    const addListener = () => true;

    const tree = renderer.create(
      <Position
        navigation={{ isFocused, addListener }}
        selectedObservation={observation}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
