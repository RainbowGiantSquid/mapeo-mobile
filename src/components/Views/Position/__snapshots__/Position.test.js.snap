// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Position tests snapshots 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#333333",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": 60,
          "paddingTop": 15,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 15,
            "marginTop": 3,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "gray",
                "fontSize": 25,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignItems": "center",
            "color": "#EAEAEA",
            "flex": 1,
            "fontSize": 20,
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Posicíon
      </Text>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#ff9933",
            "borderRadius": 15,
            "height": 30,
            "marginRight": 15,
            "opacity": 1,
            "width": 45,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 25,
              },
              Object {
                "alignSelf": "center",
                "justifyContent": "center",
                "marginTop": 2,
              },
              Object {
                "fontFamily": "Feather",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "height": 200,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "backgroundColor": "#333333",
            "borderRadius": 20,
            "color": "white",
            "fontWeight": "700",
            "marginTop": 15,
            "paddingVertical": 10,
            "textAlign": "center",
            "width": 150,
            "zIndex": 5,
          }
        }
      >
        GPS: Fuerte
      </Text>
      <Image
        source={
          Object {
            "testUri": "../../../src/images/confirm-position.png",
          }
        }
        style={
          Object {
            "alignSelf": "stretch",
            "flex": 1,
            "height": 200,
            "position": "absolute",
            "width": 750,
          }
        }
      />
      <Image
        source={
          Object {
            "testUri": "../../../src/images/position.png",
          }
        }
        style={
          Object {
            "height": 47,
            "marginTop": 20,
            "width": 40,
            "zIndex": 5,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "margin": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#EAEAEA",
            "fontSize": 12,
            "fontWeight": "700",
            "marginBottom": 10,
          }
        }
      >
        Tu posicíon
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholderTextColor="#666666"
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#1F1F1F",
            "color": "#ffffff",
            "fontSize": 20,
            "fontWeight": "700",
            "justifyContent": "flex-start",
            "marginBottom": 15,
            "padding": 20,
          }
        }
        underlineColorAndroid="#1F1F1F"
        value="50.5, 50.5"
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#EAEAEA",
            "fontSize": 12,
            "fontWeight": "700",
            "marginBottom": 10,
          }
        }
      >
        ?Qué tan lejos está la observación?
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Por ejemplo: 20 metros"
        placeholderTextColor="#666666"
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#1F1F1F",
            "color": "#ffffff",
            "fontSize": 20,
            "fontWeight": "700",
            "justifyContent": "flex-start",
            "marginBottom": 15,
            "padding": 20,
          }
        }
        underlineColorAndroid="#1F1F1F"
        value=""
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#EAEAEA",
            "fontSize": 12,
            "fontWeight": "700",
            "marginBottom": 10,
          }
        }
      >
        Cerce de...
      </Text>
      <View
        style={
          Object {
            "borderColor": "#1F1F1F",
            "borderWidth": 2,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          style={
            Object {
              "color": "#ffffff",
              "flex": 1,
              "fontSize": 20,
              "fontWeight": "700",
              "padding": 20,
            }
          }
          underlineColorAndroid="#333333"
          value="Sinangoe"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
